src/StoreApi/Routes/Checkout.php

```diff
@@ -139,71 +139,86 @@ protected function get_route_update_response( \WP_REST_Request $request ) {
 	}
 
 	/**
-	 * Update and process payment for the order.
+	 * Update and process an order.
+	 *
+	 * 1. Obtain Draft Order
+	 * 2. Process Request
+	 * 3. Process Customer
+	 * 4. Validate Order
+	 * 5. Process Payment
 	 *
 	 * @throws RouteException On error.
 	 * @param \WP_REST_Request $request Request object.
 	 * @return \WP_REST_Response
 	 */
 	protected function get_route_post_response( \WP_REST_Request $request ) {
+		// --
+		// 1. Obtain Draft Order
+
 		// Update customer first since orders will be created using that data.
 		$this->update_customer_from_request( $request );
 
-		$order_controller = new OrderController();
-		$order_object     = $this->get_draft_order_object( $this->get_draft_order_id() );
+		// Get the order, including validation and reserving stock.
+		$order = $this->create_or_update_draft_order();
 
-		if ( ! $order_object instanceof \WC_Order ) {
-			throw new RouteException(
-				'woocommerce_rest_checkout_invalid_order',
-				__( 'This session has no orders pending payment.', 'woo-gutenberg-products-block' ),
-				500
-			);
-		}
+		// --
+		// 2. Process Request
 
-		// Ensure order still matches cart.
-		$order_controller->update_order_from_cart( $order_object );
+		// If any form fields were posted, update the order (e.g. order note, payment method).
+		$this->update_order_from_request( $order, $request );
 
-		// Create a new user account as necessary.
-		// Note - CreateAccount class includes feature gating logic (i.e. this
-		// may not create an account depending on build).
-		if ( defined( 'WC_VERSION' ) && version_compare( WC_VERSION, '4.7', '>=' ) ) {
-			// Checkout signup is feature gated to WooCommerce 4.7 and newer;
-			// Because it requires updated my-account/lost-password screen in 4.7+
-			// for setting initial password.
-			try {
-				$create_account = Package::container()->get( CreateAccount::class );
-				$create_account->from_order_request( $request );
-				$order_object->set_customer_id( get_current_user_id() );
-			} catch ( Exception $error ) {
-				$this->handle_error( $error );
-			}
+		// Validate payment method before proceeding.
+		if ( $order->needs_payment() ) {
+			// Will throw and return 400 if payment method missing or invalid.
+			$this->get_request_payment_method_id( $request );
```

Comment:
This method is being used to validate, when it explicitly returns a payment method for the given id. Instead, we should maybe modify `process_payment` so it receives an optional `$payment_method_id` as an argument, then you can retrieve the payment method here and feed it into `process_payment`. Then `process_payment` only needs to call `get_request_payment_method_id` if it's not present.

So `Checkout::process_payment` would have this as its signature:

```php
protected function process_payment( \WC_Order $order, \WP_REST_Request $request, string $payment_method_id  = '' ) { 
    /** ... **/
   $context->set_payment_method( $payment_method_id === '' ? $this->get_request_payment_method_id( $request ) : $payment_method_id );
    /** ... **/
}
```

Reply:
I think we can also move validation to the schema (enum property)